risc0/zkvm/src/host/server/session.rs

```diff
@@ -135,6 +153,76 @@ impl Session {
         self.hooks.push(Box::new(hook));
     }
 
+    /// Calculate for the [ReceiptMetadata] associated with this [Session]. The
+    /// [ReceiptMetadata] is the claim that will be proven if this [Session]
+    /// is passed to the [Prover].
+    pub fn get_metadata(&self) -> Result<ReceiptMetadata> {
+        let first_segment = &self
+            .segments
+            .first()
+            .ok_or_else(|| anyhow!("session has no segments"))?
+            .resolve()?;
+        let last_segment = &self
+            .segments
+            .last()
+            .ok_or_else(|| anyhow!("session has no segments"))?
+            .resolve()?;
+
+        // Construct the Output struct, checking that the Session is internally
+        // consistent.
+        let output = if self.exit_code.expects_output() {
+            self.journal
+                .as_ref()
+                .map(|journal| -> Result<_> {
+                    Ok(Output {
+                        journal: journal.clone().into(),
+                        assumptions: Assumptions(
+                            self.assumptions
+                                .iter()
+                                .filter_map(|a| match a {
+                                    Assumption::Proven(_) => None,
+                                    Assumption::Unresolved(r) => Some(r.clone()),
+                                })
+                                .collect::<Vec<_>>(),
+                        )
+                        .into(),
+                    })
+                })
+                .transpose()?
+        } else {
+            ensure!(
+                self.journal.is_none(),
+                "Session with exit code {:?} has a journal",
+                self.exit_code
+            );
+            ensure!(
+                self.assumptions.is_empty(),
+                "Session with exit code {:?} has encoded assumptions",
+                self.exit_code
+            );
+            None
+        };
+
+        // DO NOT MERGE: post should match the final segment system state, but doesn't
+        // always. Prior to merging figure out what is happening here and what
+        // to do about it (possibly nothing).
```

Comment:
Should we keep an open comment in this PR to track this?

Reply:
Yes, thank you